### ISSUE TITLE
Replace Pyomo

### ISSUE THREAD

**Issue by greyhillman** [hues-platform/python-ehub#47]
The code for Pyomo is an absolute mess. There is no consistency with accessing stuff, it is not Pythonic at all, and sucks for working with in an IDE like PyCharm.

It will be a lot of work, but hopefully it will make more sense than what is currently being done behind the scenes.

**Comment by greyhillman** [comment]
For our use case, we just want something that will run on our local machines. It should be really simple to use and understand.
It should not use `*args` and `**kwargs` for every method and function, that's just rude. It makes it impossible to use PyCharm to find out what the arguments and keyword arguments are.

Also, there should be docstrings readily accessible by IDEs like PyCharm. I should be able to hit `Ctrl+q` and see what the arguments are for and what the method does.

The naming conventions should be similar. I shouldn't have to call `get_values()`, `extract_values()` and `values` for Pyomo Variables, Parameters, and Sets just to actually get the underlying dictionary representation. All of them should be a simply property like `values`.

I would also think it's a good idea to try and use Panda's dataframes or Numpy matricies as the underlying data holder. As of writing, Panda's DataFrames are used to output the data in a nice format in the terminal and output file. I would like to be able to just have that ready to go as easily as possible. I think it uses `DataFrame.from_dict` to convert from the response format into the DataFrame.

Pyomo is also a prime example of the overuse of OOP. A `Param` shouldn't have more than like 3 ancestors. Pyomo gives it like 10 or more.

The structure of the library should be as close to a pure mathematical linear programming way as possible. You should have something like a `Variable` class for variables, `Constant` for all other numbers, and `Constraint` for a constraint. When creating a program, have the user pass in the objective function and a list of all the constraints themselves. That makes the solver much easier to use. When the solver has solved the problem, the library can send back a dictionary of the names of the variables and their values.

**Comment by greyhillman** [comment]
The Python library PuLP seems like it can be used as a basis for a cleaner wrapper.
Here's an example of it: http://benalexkeen.com/linear-programming-with-python-and-pulp-part-5/

Google has their own as well: https://developers.google.com/optimization/mip/integer_opt

Most of them are not really Pythonic though. Might still have to write a wrapper.

**Comment by ralphevins** [comment]
Can you estimate how much work it would be to build a simple wrapper along the pure-LP, very Pythonic lines that you discuss?

**Comment by greyhillman** [comment]
I've been working on that this morning. I think I have almost everything actually. It's on the `replace_pyomo` branch. You can look at the `pylp_test.py` file for an example.

**Comment by ralphevins** [comment]
A critical thing we found previously is how well the wrapper handles large arrays of variables i.e. for 8760 time steps. Would be good to see a test of this with your wrapper.

**Comment by greyhillman** [comment]
Thanks for telling me this. I ended up getting some 'max recursion' errors and I'm now working on fixing that. Hopefully the underlying library can handle it (it seems like it can).

**Comment by greyhillman** [comment]
I've fixed the problem and now have it working on the full 8760 time series model.

The only big difference is that there is some preprocessing being performed by GLPK before solving. I'm guessing this has to do with some stuff like simplifying constants (like right now if you have `5 * x * 2` it leaves it like that instead of converting to `10 * x` or `x * 10`).

While writing this, it solved the model I had for it. It must have been because we got rid of the variables for capacities. All in all, it looks good.
I can even have the constraint printed out in a nice format as well, which is something Pyomo could never do.
